run/jenkins.py

```diff
@@ -0,0 +1,282 @@
+# This replaces run/run.py
+import json
+import gzip
+import os
+import requests
+import subprocess
+import time
+import re
+
+
+METADATA_URL = 'http://metadata.google.internal/computeMetadata/v1'
+
+
+def main():
+    # TODO after --install-browser verify browser
+    # version or change platform ID to correct browser version
+    # This can be done by running `wpt install firefox` and then verifying
+    # TODO report OS version when creating TestRun
+    # TODO check out correct WPT revision
+
+    # TODO convert these back into command line args. They're currently
+    # env vars due to an earlier change. They should be converted back.
+
+    # TODO break this file up into smaller, unit tested modules.
+    args = {
+        'prod_run': bool(os.environ.get('PROD_RUN', False)),
+        'prod_wet_run': bool(os.environ.get('PROD_WET_RUN', False)),
+        'SHA': os.environ.get('WPT_SHA'),
+        'sauce_from_metadata': bool(os.environ.get('SAUCE_FROM_METADATA')),
+        'sauce_key': os.environ.get('SAUCE_KEY', ''),
+        'sauce_user': os.environ.get('SAUCE_USER', ''),
+        'jenkins_job_name': os.environ.get('JOB_NAME'),
+        'wpt_path': os.environ.get('WPT_PATH'),
+        'wptd_path': '/wptdashboard',
+        'output_path': '/wptdout',
+        'run_path': os.environ.get('RUN_PATH', ''),
+    }
+
+    # Arg validation
+    assert args['wpt_path'], '`WPT_PATH` env var required.'
+    assert not args['wpt_path'].endswith('/'), '`WPT_PATH` cannot end with /.'
+
+    if args['prod_run'] or args['prod_wet_run']:
+        print 'Fetching upload_secret from GCE Metadata...'
+
+        url = '%s/project/attributes/upload_secret' % METADATA_URL
+        res = requests.get(url, headers={'Metadata-Flavor': 'Google'})
+        args['upload_secret'] = res.text
+        assert len(args['upload_secret']) == 64, (
+            'Metadata `upload_secret` must exist for prod runs.')
```

Comment:
Is there a common reason this would occur? Is it usually failing to setup something (like app default creds on a gcloud project)? If so, maybe we could have a more informative message.

Reply:
I haven't yet seen this happen. I'll make a note to document how we use metadata somewhere.